src/FESpaces/FESpacesWithDofPotentiallyRemoved.jl

```diff
@@ -0,0 +1,185 @@
+    @notimplementedif num_dirichlet_dofs(space) != 0 s
+    @assert !remove_dof || 1 <= dof_to_remove <= num_free_dofs(space)
+    cs = constraint_style(space)
+    CS = get_val_parameter(cs)
+    rd = Val{remove_dof}()
+    RD = remove_dof
+    new{CS,RD}(space,cs,rd,dof_to_remove)
+  end
+end
+
+# Genuine functions
+function num_free_dofs(f::FESpaceWithDofPotentiallyRemoved{CS,true}) where {CS}
+  num_free_dofs(f.space)-1
+end
+
+function num_free_dofs(f::FESpaceWithDofPotentiallyRemoved{CS,false}) where {CS}
+  num_free_dofs(f.space)
+end
+
+function zero_free_values(f::FESpaceWithDofPotentiallyRemoved)
+  zeros(num_free_dofs(f))
+end
+
+function get_cell_dofs(f::FESpaceWithDofPotentiallyRemoved{CS,true}) where {CS}
+  cell_dofs = get_cell_dofs(f.space)
+  CellDofsWithDofRemoved(cell_dofs,f.dof_to_remove)
+end
+
+function get_cell_dofs(f::FESpaceWithDofPotentiallyRemoved{CS,false}) where {CS}
+  get_cell_dofs(f.space)
+end
+
+
+num_dirichlet_dofs(f::FESpaceWithDofPotentiallyRemoved{CS,true}) where {CS} = 1
+
+num_dirichlet_dofs(f::FESpaceWithDofPotentiallyRemoved{CS,false}) where {CS} = 0
+
+function zero_dirichlet_values(f::FESpaceWithDofPotentiallyRemoved)
+  T = Float64 # TODO
+  zeros(T,num_dirichlet_dofs(f))
+end
+
+
+num_dirichlet_tags(f::FESpaceWithDofPotentiallyRemoved{CS,true}) where {CS} = 1
+
+num_dirichlet_tags(f::FESpaceWithDofPotentiallyRemoved{CS,false}) where {CS} = 0
+
+get_dirichlet_dof_tag(f::FESpaceWithDofPotentiallyRemoved{CS,true}) where {CS} = Int8[1,]
+
+get_dirichlet_dof_tag(f::FESpaceWithDofPotentiallyRemoved{CS,false}) where {CS} = Int8[]
+
+function scatter_free_and_dirichlet_values(
+  f::FESpaceWithDofPotentiallyRemoved{CS,true},fv,dv) where {CS}
+  @assert length(dv) == 1
+  _dv = similar(dv,eltype(dv),0)
+  _fv = vcat(view(fv,1:f.dof_to_remove-1),
+             dv,
+             view(fv,f.dof_to_remove:length(fv))) # TODO lazy append
+ scatter_free_and_dirichlet_values(f.space,_fv,_dv)
+end
+
+function scatter_free_and_dirichlet_values(
+  f::FESpaceWithDofPotentiallyRemoved{CS,false},fv,dv) where {CS}
+  @assert length(dv) == 0
+ scatter_free_and_dirichlet_values(f.space,fv,dv)
+end
+
+function gather_free_and_dirichlet_values(
+  f::FESpaceWithDofPotentiallyRemoved{CS,true},cv) where {CS}
+  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
+  @assert length(_dv) == 0
+  fv = vcat(view(_fv,1:f.dof_to_remove-1),
+            view(_fv,f.dof_to_remove+1:length(_fv))) # TODO: can we avoid
+                                          # allocating new memory?
+  dv = view(_fv,f.dof_to_remove:f.dof_to_remove)
+  (fv, dv)
+end
+
+function gather_free_and_dirichlet_values(
+  f::FESpaceWithDofPotentiallyRemoved{CS,false},cv) where {CS}
+  gather_free_and_dirichlet_values(f.space,cv)
+end
+
+function gather_free_and_dirichlet_values!(
+  fv,dv,f::FESpaceWithDofPotentiallyRemoved{CS,true},cv) where {CS}
+  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
```

Comment:
Can we reduce code duplication between `gather_free_and_dirichlet_values` and `gather_free_and_dirichlet_values!` ?

Reply:
> Can we reduce code duplication between gather_free_and_dirichlet_values and gather_free_and_dirichlet_values! ?

I guess the comment has become outdated after latest changes. Agreed? (See below)

```julia
function gather_free_and_dirichlet_values(
  f::FESpaceWithConstantFixed{CS,FixConstant},cv) where {CS}
  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
  @assert length(_dv) == 0
  fv = VectorWithEntryRemoved(_fv,f.dof_to_fix)
  dv = _fv[f.dof_to_fix:f.dof_to_fix]
  (fv, dv)
end

function gather_free_and_dirichlet_values!(
  fv,dv,f::FESpaceWithConstantFixed{CS,FixConstant},cv) where {CS}
  _fv, _dv = gather_free_and_dirichlet_values(f.space,cv)
  @assert length(_dv) == 0
  fv    .= VectorWithEntryRemoved(_fv,f.dof_to_fix)
  dv[1]  = _fv[f.dof_to_fix]
  (fv, dv)
end
```